{"name": "Успешная регистрация студента с полными данными", "status": "broken", "statusDetails": {"message": "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=59269): Max retries exceeded with url: /session/94e97799e125329f558cf44c1e7b7b38/screenshot (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2f00>: Failed to establish a new connection: [Errno 61] Connection refused'))", "trace": "self = <urllib3.connection.HTTPConnection object at 0x107fe2120>\n\n    def _new_conn(self) -> socket.socket:\n        \"\"\"Establish a socket connection and set nodelay settings on it.\n    \n        :return: New socket connection.\n        \"\"\"\n        try:\n>           sock = connection.create_connection(\n                (self._dns_host, self.port),\n                self.timeout,\n                source_address=self.source_address,\n                socket_options=self.socket_options,\n            )\n\n.venv/lib/python3.12/site-packages/urllib3/connection.py:198: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n.venv/lib/python3.12/site-packages/urllib3/util/connection.py:85: in create_connection\n    raise err\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\naddress = ('localhost', 59269), timeout = 120, source_address = None, socket_options = [(6, 1, 1)]\n\n    def create_connection(\n        address: tuple[str, int],\n        timeout: _TYPE_TIMEOUT = _DEFAULT_TIMEOUT,\n        source_address: tuple[str, int] | None = None,\n        socket_options: _TYPE_SOCKET_OPTIONS | None = None,\n    ) -> socket.socket:\n        \"\"\"Connect to *address* and return the socket object.\n    \n        Convenience function.  Connect to *address* (a 2-tuple ``(host,\n        port)``) and return the socket object.  Passing the optional\n        *timeout* parameter will set the timeout on the socket instance\n        before attempting to connect.  If no *timeout* is supplied, the\n        global default timeout setting returned by :func:`socket.getdefaulttimeout`\n        is used.  If *source_address* is set it must be a tuple of (host, port)\n        for the socket to bind as a source address before making the connection.\n        An host of '' or port 0 tells the OS to use the default.\n        \"\"\"\n    \n        host, port = address\n        if host.startswith(\"[\"):\n            host = host.strip(\"[]\")\n        err = None\n    \n        # Using the value from allowed_gai_family() in the context of getaddrinfo lets\n        # us select whether to work with IPv4 DNS records, IPv6 records, or both.\n        # The original create_connection function always returns all records.\n        family = allowed_gai_family()\n    \n        try:\n            host.encode(\"idna\")\n        except UnicodeError:\n            raise LocationParseError(f\"'{host}', label empty or too long\") from None\n    \n        for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n            af, socktype, proto, canonname, sa = res\n            sock = None\n            try:\n                sock = socket.socket(af, socktype, proto)\n    \n                # If provided, set socket level options before connecting.\n                _set_socket_options(sock, socket_options)\n    \n                if timeout is not _DEFAULT_TIMEOUT:\n                    sock.settimeout(timeout)\n                if source_address:\n                    sock.bind(source_address)\n>               sock.connect(sa)\nE               ConnectionRefusedError: [Errno 61] Connection refused\n\n.venv/lib/python3.12/site-packages/urllib3/util/connection.py:73: ConnectionRefusedError\n\nThe above exception was the direct cause of the following exception:\n\nself = <urllib3.connectionpool.HTTPConnectionPool object at 0x102f72300>, method = 'POST', url = '/session/94e97799e125329f558cf44c1e7b7b38/element'\nbody = '{\"using\": \"css selector\", \"value\": \".table\"}'\nheaders = HTTPHeaderDict({'Accept': 'application/json', 'Content-Type': 'application/json;charset=UTF-8', 'User-Agent': 'selenium/4.34.0 (python mac)', 'Connection': 'keep-alive'})\nretries = Retry(total=0, connect=None, read=None, redirect=None, status=None), redirect = False, assert_same_host = False, timeout = 120, pool_timeout = None\nrelease_conn = True, chunked = False, body_pos = None, preload_content = True, decode_content = True, response_kw = {}\nparsed_url = Url(scheme=None, auth=None, host=None, port=None, path='/session/94e97799e125329f558cf44c1e7b7b38/element', query=None, fragment=None)\ndestination_scheme = None, conn = None, release_this_conn = True, http_tunnel_required = False, err = None, clean_exit = False\n\n    def urlopen(  # type: ignore[override]\n        self,\n        method: str,\n        url: str,\n        body: _TYPE_BODY | None = None,\n        headers: typing.Mapping[str, str] | None = None,\n        retries: Retry | bool | int | None = None,\n        redirect: bool = True,\n        assert_same_host: bool = True,\n        timeout: _TYPE_TIMEOUT = _DEFAULT_TIMEOUT,\n        pool_timeout: int | None = None,\n        release_conn: bool | None = None,\n        chunked: bool = False,\n        body_pos: _TYPE_BODY_POSITION | None = None,\n        preload_content: bool = True,\n        decode_content: bool = True,\n        **response_kw: typing.Any,\n    ) -> BaseHTTPResponse:\n        \"\"\"\n        Get a connection from the pool and perform an HTTP request. This is the\n        lowest level call for making a request, so you'll need to specify all\n        the raw details.\n    \n        .. note::\n    \n           More commonly, it's appropriate to use a convenience method\n           such as :meth:`request`.\n    \n        .. note::\n    \n           `release_conn` will only behave as expected if\n           `preload_content=False` because we want to make\n           `preload_content=False` the default behaviour someday soon without\n           breaking backwards compatibility.\n    \n        :param method:\n            HTTP request method (such as GET, POST, PUT, etc.)\n    \n        :param url:\n            The URL to perform the request on.\n    \n        :param body:\n            Data to send in the request body, either :class:`str`, :class:`bytes`,\n            an iterable of :class:`str`/:class:`bytes`, or a file-like object.\n    \n        :param headers:\n            Dictionary of custom headers to send, such as User-Agent,\n            If-None-Match, etc. If None, pool headers are used. If provided,\n            these headers completely replace any pool-specific headers.\n    \n        :param retries:\n            Configure the number of retries to allow before raising a\n            :class:`~urllib3.exceptions.MaxRetryError` exception.\n    \n            If ``None`` (default) will retry 3 times, see ``Retry.DEFAULT``. Pass a\n            :class:`~urllib3.util.retry.Retry` object for fine-grained control\n            over different types of retries.\n            Pass an integer number to retry connection errors that many times,\n            but no other types of errors. Pass zero to never retry.\n    \n            If ``False``, then retries are disabled and any exception is raised\n            immediately. Also, instead of raising a MaxRetryError on redirects,\n            the redirect response will be returned.\n    \n        :type retries: :class:`~urllib3.util.retry.Retry`, False, or an int.\n    \n        :param redirect:\n            If True, automatically handle redirects (status codes 301, 302,\n            303, 307, 308). Each redirect counts as a retry. Disabling retries\n            will disable redirect, too.\n    \n        :param assert_same_host:\n            If ``True``, will make sure that the host of the pool requests is\n            consistent else will raise HostChangedError. When ``False``, you can\n            use the pool on an HTTP proxy and request foreign hosts.\n    \n        :param timeout:\n            If specified, overrides the default timeout for this one\n            request. It may be a float (in seconds) or an instance of\n            :class:`urllib3.util.Timeout`.\n    \n        :param pool_timeout:\n            If set and the pool is set to block=True, then this method will\n            block for ``pool_timeout`` seconds and raise EmptyPoolError if no\n            connection is available within the time period.\n    \n        :param bool preload_content:\n            If True, the response's body will be preloaded into memory.\n    \n        :param bool decode_content:\n            If True, will attempt to decode the body based on the\n            'content-encoding' header.\n    \n        :param release_conn:\n            If False, then the urlopen call will not release the connection\n            back into the pool once a response is received (but will release if\n            you read the entire contents of the response such as when\n            `preload_content=True`). This is useful if you're not preloading\n            the response's content immediately. You will need to call\n            ``r.release_conn()`` on the response ``r`` to return the connection\n            back into the pool. If None, it takes the value of ``preload_content``\n            which defaults to ``True``.\n    \n        :param bool chunked:\n            If True, urllib3 will send the body using chunked transfer\n            encoding. Otherwise, urllib3 will send the body using the standard\n            content-length form. Defaults to False.\n    \n        :param int body_pos:\n            Position to seek to in file-like body in the event of a retry or\n            redirect. Typically this won't need to be set because urllib3 will\n            auto-populate the value when needed.\n        \"\"\"\n        parsed_url = parse_url(url)\n        destination_scheme = parsed_url.scheme\n    \n        if headers is None:\n            headers = self.headers\n    \n        if not isinstance(retries, Retry):\n            retries = Retry.from_int(retries, redirect=redirect, default=self.retries)\n    \n        if release_conn is None:\n            release_conn = preload_content\n    \n        # Check host\n        if assert_same_host and not self.is_same_host(url):\n            raise HostChangedError(self, url, retries)\n    \n        # Ensure that the URL we're connecting to is properly encoded\n        if url.startswith(\"/\"):\n            url = to_str(_encode_target(url))\n        else:\n            url = to_str(parsed_url.url)\n    \n        conn = None\n    \n        # Track whether `conn` needs to be released before\n        # returning/raising/recursing. Update this variable if necessary, and\n        # leave `release_conn` constant throughout the function. That way, if\n        # the function recurses, the original value of `release_conn` will be\n        # passed down into the recursive call, and its value will be respected.\n        #\n        # See issue #651 [1] for details.\n        #\n        # [1] <https://github.com/urllib3/urllib3/issues/651>\n        release_this_conn = release_conn\n    \n        http_tunnel_required = connection_requires_http_tunnel(\n            self.proxy, self.proxy_config, destination_scheme\n        )\n    \n        # Merge the proxy headers. Only done when not using HTTP CONNECT. We\n        # have to copy the headers dict so we can safely change it without those\n        # changes being reflected in anyone else's copy.\n        if not http_tunnel_required:\n            headers = headers.copy()  # type: ignore[attr-defined]\n            headers.update(self.proxy_headers)  # type: ignore[union-attr]\n    \n        # Must keep the exception bound to a separate variable or else Python 3\n        # complains about UnboundLocalError.\n        err = None\n    \n        # Keep track of whether we cleanly exited the except block. This\n        # ensures we do proper cleanup in finally.\n        clean_exit = False\n    \n        # Rewind body position, if needed. Record current position\n        # for future rewinds in the event of a redirect/retry.\n        body_pos = set_file_position(body, body_pos)\n    \n        try:\n            # Request a connection from the queue.\n            timeout_obj = self._get_timeout(timeout)\n            conn = self._get_conn(timeout=pool_timeout)\n    \n            conn.timeout = timeout_obj.connect_timeout  # type: ignore[assignment]\n    \n            # Is this a closed/new connection that requires CONNECT tunnelling?\n            if self.proxy is not None and http_tunnel_required and conn.is_closed:\n                try:\n                    self._prepare_proxy(conn)\n                except (BaseSSLError, OSError, SocketTimeout) as e:\n                    self._raise_timeout(\n                        err=e, url=self.proxy.url, timeout_value=conn.timeout\n                    )\n                    raise\n    \n            # If we're going to release the connection in ``finally:``, then\n            # the response doesn't need to know about the connection. Otherwise\n            # it will also try to release it and we'll have a double-release\n            # mess.\n            response_conn = conn if not release_conn else None\n    \n            # Make the request on the HTTPConnection object\n>           response = self._make_request(\n                conn,\n                method,\n                url,\n                timeout=timeout_obj,\n                body=body,\n                headers=headers,\n                chunked=chunked,\n                retries=retries,\n                response_conn=response_conn,\n                preload_content=preload_content,\n                decode_content=decode_content,\n                **response_kw,\n            )\n\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:787: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:493: in _make_request\n    conn.request(\n.venv/lib/python3.12/site-packages/urllib3/connection.py:445: in request\n    self.endheaders()\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1314: in endheaders\n    self._send_output(message_body, encode_chunked=encode_chunked)\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1074: in _send_output\n    self.send(msg)\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1018: in send\n    self.connect()\n.venv/lib/python3.12/site-packages/urllib3/connection.py:276: in connect\n    self.sock = self._new_conn()\n                ^^^^^^^^^^^^^^^^\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\nself = <urllib3.connection.HTTPConnection object at 0x107fe2120>\n\n    def _new_conn(self) -> socket.socket:\n        \"\"\"Establish a socket connection and set nodelay settings on it.\n    \n        :return: New socket connection.\n        \"\"\"\n        try:\n            sock = connection.create_connection(\n                (self._dns_host, self.port),\n                self.timeout,\n                source_address=self.source_address,\n                socket_options=self.socket_options,\n            )\n        except socket.gaierror as e:\n            raise NameResolutionError(self.host, self, e) from e\n        except SocketTimeout as e:\n            raise ConnectTimeoutError(\n                self,\n                f\"Connection to {self.host} timed out. (connect timeout={self.timeout})\",\n            ) from e\n    \n        except OSError as e:\n>           raise NewConnectionError(\n                self, f\"Failed to establish a new connection: {e}\"\n            ) from e\nE           urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x107fe2120>: Failed to establish a new connection: [Errno 61] Connection refused\n\n.venv/lib/python3.12/site-packages/urllib3/connection.py:213: NewConnectionError\n\nThe above exception was the direct cause of the following exception:\n\nfn = <selene.core.conditions.CollectionCondition object at 0x102f737d0>\n\n    def logic(fn: Callable[[E], R]) -> R:\n        finish_time = time.time() + self._timeout\n    \n        while True:\n            try:\n>               return fn(self.entity)\n                       ^^^^^^^^^^^^^^^\n\n.venv/lib/python3.12/site-packages/selene/core/wait.py:119: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n.venv/lib/python3.12/site-packages/selene/core/condition.py:174: in __call__\n    return self._fn(entity)\n           ^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/condition.py:144: in fn\n    actual = query(entity)\n             ^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/match.py:389: in visible_texts\n    webelement.text for webelement in collection() if webelement.is_displayed()\n                                      ^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:648: in __call__\n    return self.locate()\n           ^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:641: in locate\n    return self._locator()\n           ^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/locator.py:34: in __call__\n    return self._locate()\n           ^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:719: in find\n    webelements = self()\n                  ^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:648: in __call__\n    return self.locate()\n           ^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:641: in locate\n    return self._locator()\n           ^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/locator.py:34: in __call__\n    return self._locate()\n           ^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:249: in <lambda>\n    Locator(f'{self}.all({by})', lambda: self().find_elements(*by)),\n                                         ^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:192: in __call__\n    return self.locate()\n           ^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/entity.py:185: in locate\n    return self._locator()\n           ^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/locator.py:34: in __call__\n    return self._locate()\n           ^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/_browser.py:71: in <lambda>\n    Locator(f'{self}.element({by})', lambda: self.driver.find_element(*by)),\n                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:922: in find_element\n    return self.execute(Command.FIND_ELEMENT, {\"using\": by, \"value\": value})[\"value\"]\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:451: in execute\n    response = cast(RemoteConnection, self.command_executor).execute(driver_command, params)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py:404: in execute\n    return self._request(command_info[0], url, body=data)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py:428: in _request\n    response = self._conn.request(method, url, body=body, headers=headers, timeout=self._client_config.timeout)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/_request_methods.py:143: in request\n    return self.request_encode_body(\n.venv/lib/python3.12/site-packages/urllib3/_request_methods.py:278: in request_encode_body\n    return self.urlopen(method, url, **extra_kw)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/poolmanager.py:443: in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:841: in urlopen\n    retries = retries.increment(\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\nself = Retry(total=0, connect=None, read=None, redirect=None, status=None), method = 'POST', url = '/session/94e97799e125329f558cf44c1e7b7b38/element'\nresponse = None\nerror = NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2120>: Failed to establish a new connection: [Errno 61] Connection refused')\n_pool = <urllib3.connectionpool.HTTPConnectionPool object at 0x102f72300>, _stacktrace = <traceback object at 0x107ffb840>\n\n    def increment(\n        self,\n        method: str | None = None,\n        url: str | None = None,\n        response: BaseHTTPResponse | None = None,\n        error: Exception | None = None,\n        _pool: ConnectionPool | None = None,\n        _stacktrace: TracebackType | None = None,\n    ) -> Self:\n        \"\"\"Return a new Retry object with incremented retry counters.\n    \n        :param response: A response object, or None, if the server did not\n            return a response.\n        :type response: :class:`~urllib3.response.BaseHTTPResponse`\n        :param Exception error: An error encountered during the request, or\n            None if the response was received successfully.\n    \n        :return: A new ``Retry`` object.\n        \"\"\"\n        if self.total is False and error:\n            # Disabled, indicate to re-raise the error.\n            raise reraise(type(error), error, _stacktrace)\n    \n        total = self.total\n        if total is not None:\n            total -= 1\n    \n        connect = self.connect\n        read = self.read\n        redirect = self.redirect\n        status_count = self.status\n        other = self.other\n        cause = \"unknown\"\n        status = None\n        redirect_location = None\n    \n        if error and self._is_connection_error(error):\n            # Connect retry?\n            if connect is False:\n                raise reraise(type(error), error, _stacktrace)\n            elif connect is not None:\n                connect -= 1\n    \n        elif error and self._is_read_error(error):\n            # Read retry?\n            if read is False or method is None or not self._is_method_retryable(method):\n                raise reraise(type(error), error, _stacktrace)\n            elif read is not None:\n                read -= 1\n    \n        elif error:\n            # Other retry?\n            if other is not None:\n                other -= 1\n    \n        elif response and response.get_redirect_location():\n            # Redirect retry?\n            if redirect is not None:\n                redirect -= 1\n            cause = \"too many redirects\"\n            response_redirect_location = response.get_redirect_location()\n            if response_redirect_location:\n                redirect_location = response_redirect_location\n            status = response.status\n    \n        else:\n            # Incrementing because of a server error like a 500 in\n            # status_forcelist and the given method is in the allowed_methods\n            cause = ResponseError.GENERIC_ERROR\n            if response and response.status:\n                if status_count is not None:\n                    status_count -= 1\n                cause = ResponseError.SPECIFIC_ERROR.format(status_code=response.status)\n                status = response.status\n    \n        history = self.history + (\n            RequestHistory(method, url, error, status, redirect_location),\n        )\n    \n        new_retry = self.new(\n            total=total,\n            connect=connect,\n            read=read,\n            redirect=redirect,\n            status=status_count,\n            other=other,\n            history=history,\n        )\n    \n        if new_retry.is_exhausted():\n            reason = error or ResponseError(cause)\n>           raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]\n            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\nE           urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=59269): Max retries exceeded with url: /session/94e97799e125329f558cf44c1e7b7b38/element (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2120>: Failed to establish a new connection: [Errno 61] Connection refused'))\n\n.venv/lib/python3.12/site-packages/urllib3/util/retry.py:519: MaxRetryError\n\nDuring handling of the above exception, another exception occurred:\n\nself = <urllib3.connection.HTTPConnection object at 0x107fe2f00>\n\n    def _new_conn(self) -> socket.socket:\n        \"\"\"Establish a socket connection and set nodelay settings on it.\n    \n        :return: New socket connection.\n        \"\"\"\n        try:\n>           sock = connection.create_connection(\n                (self._dns_host, self.port),\n                self.timeout,\n                source_address=self.source_address,\n                socket_options=self.socket_options,\n            )\n\n.venv/lib/python3.12/site-packages/urllib3/connection.py:198: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n.venv/lib/python3.12/site-packages/urllib3/util/connection.py:85: in create_connection\n    raise err\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\naddress = ('localhost', 59269), timeout = 120, source_address = None, socket_options = [(6, 1, 1)]\n\n    def create_connection(\n        address: tuple[str, int],\n        timeout: _TYPE_TIMEOUT = _DEFAULT_TIMEOUT,\n        source_address: tuple[str, int] | None = None,\n        socket_options: _TYPE_SOCKET_OPTIONS | None = None,\n    ) -> socket.socket:\n        \"\"\"Connect to *address* and return the socket object.\n    \n        Convenience function.  Connect to *address* (a 2-tuple ``(host,\n        port)``) and return the socket object.  Passing the optional\n        *timeout* parameter will set the timeout on the socket instance\n        before attempting to connect.  If no *timeout* is supplied, the\n        global default timeout setting returned by :func:`socket.getdefaulttimeout`\n        is used.  If *source_address* is set it must be a tuple of (host, port)\n        for the socket to bind as a source address before making the connection.\n        An host of '' or port 0 tells the OS to use the default.\n        \"\"\"\n    \n        host, port = address\n        if host.startswith(\"[\"):\n            host = host.strip(\"[]\")\n        err = None\n    \n        # Using the value from allowed_gai_family() in the context of getaddrinfo lets\n        # us select whether to work with IPv4 DNS records, IPv6 records, or both.\n        # The original create_connection function always returns all records.\n        family = allowed_gai_family()\n    \n        try:\n            host.encode(\"idna\")\n        except UnicodeError:\n            raise LocationParseError(f\"'{host}', label empty or too long\") from None\n    \n        for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n            af, socktype, proto, canonname, sa = res\n            sock = None\n            try:\n                sock = socket.socket(af, socktype, proto)\n    \n                # If provided, set socket level options before connecting.\n                _set_socket_options(sock, socket_options)\n    \n                if timeout is not _DEFAULT_TIMEOUT:\n                    sock.settimeout(timeout)\n                if source_address:\n                    sock.bind(source_address)\n>               sock.connect(sa)\nE               ConnectionRefusedError: [Errno 61] Connection refused\n\n.venv/lib/python3.12/site-packages/urllib3/util/connection.py:73: ConnectionRefusedError\n\nThe above exception was the direct cause of the following exception:\n\nself = <urllib3.connectionpool.HTTPConnectionPool object at 0x102f72300>, method = 'GET', url = '/session/94e97799e125329f558cf44c1e7b7b38/screenshot'\nbody = None\nheaders = {'Accept': 'application/json', 'Connection': 'keep-alive', 'Content-Type': 'application/json;charset=UTF-8', 'User-Agent': 'selenium/4.34.0 (python mac)'}\nretries = Retry(total=0, connect=None, read=None, redirect=None, status=None), redirect = False, assert_same_host = False, timeout = 120, pool_timeout = None\nrelease_conn = True, chunked = False, body_pos = None, preload_content = True, decode_content = True, response_kw = {}\nparsed_url = Url(scheme=None, auth=None, host=None, port=None, path='/session/94e97799e125329f558cf44c1e7b7b38/screenshot', query=None, fragment=None)\ndestination_scheme = None, conn = None, release_this_conn = True, http_tunnel_required = False, err = None, clean_exit = False\n\n    def urlopen(  # type: ignore[override]\n        self,\n        method: str,\n        url: str,\n        body: _TYPE_BODY | None = None,\n        headers: typing.Mapping[str, str] | None = None,\n        retries: Retry | bool | int | None = None,\n        redirect: bool = True,\n        assert_same_host: bool = True,\n        timeout: _TYPE_TIMEOUT = _DEFAULT_TIMEOUT,\n        pool_timeout: int | None = None,\n        release_conn: bool | None = None,\n        chunked: bool = False,\n        body_pos: _TYPE_BODY_POSITION | None = None,\n        preload_content: bool = True,\n        decode_content: bool = True,\n        **response_kw: typing.Any,\n    ) -> BaseHTTPResponse:\n        \"\"\"\n        Get a connection from the pool and perform an HTTP request. This is the\n        lowest level call for making a request, so you'll need to specify all\n        the raw details.\n    \n        .. note::\n    \n           More commonly, it's appropriate to use a convenience method\n           such as :meth:`request`.\n    \n        .. note::\n    \n           `release_conn` will only behave as expected if\n           `preload_content=False` because we want to make\n           `preload_content=False` the default behaviour someday soon without\n           breaking backwards compatibility.\n    \n        :param method:\n            HTTP request method (such as GET, POST, PUT, etc.)\n    \n        :param url:\n            The URL to perform the request on.\n    \n        :param body:\n            Data to send in the request body, either :class:`str`, :class:`bytes`,\n            an iterable of :class:`str`/:class:`bytes`, or a file-like object.\n    \n        :param headers:\n            Dictionary of custom headers to send, such as User-Agent,\n            If-None-Match, etc. If None, pool headers are used. If provided,\n            these headers completely replace any pool-specific headers.\n    \n        :param retries:\n            Configure the number of retries to allow before raising a\n            :class:`~urllib3.exceptions.MaxRetryError` exception.\n    \n            If ``None`` (default) will retry 3 times, see ``Retry.DEFAULT``. Pass a\n            :class:`~urllib3.util.retry.Retry` object for fine-grained control\n            over different types of retries.\n            Pass an integer number to retry connection errors that many times,\n            but no other types of errors. Pass zero to never retry.\n    \n            If ``False``, then retries are disabled and any exception is raised\n            immediately. Also, instead of raising a MaxRetryError on redirects,\n            the redirect response will be returned.\n    \n        :type retries: :class:`~urllib3.util.retry.Retry`, False, or an int.\n    \n        :param redirect:\n            If True, automatically handle redirects (status codes 301, 302,\n            303, 307, 308). Each redirect counts as a retry. Disabling retries\n            will disable redirect, too.\n    \n        :param assert_same_host:\n            If ``True``, will make sure that the host of the pool requests is\n            consistent else will raise HostChangedError. When ``False``, you can\n            use the pool on an HTTP proxy and request foreign hosts.\n    \n        :param timeout:\n            If specified, overrides the default timeout for this one\n            request. It may be a float (in seconds) or an instance of\n            :class:`urllib3.util.Timeout`.\n    \n        :param pool_timeout:\n            If set and the pool is set to block=True, then this method will\n            block for ``pool_timeout`` seconds and raise EmptyPoolError if no\n            connection is available within the time period.\n    \n        :param bool preload_content:\n            If True, the response's body will be preloaded into memory.\n    \n        :param bool decode_content:\n            If True, will attempt to decode the body based on the\n            'content-encoding' header.\n    \n        :param release_conn:\n            If False, then the urlopen call will not release the connection\n            back into the pool once a response is received (but will release if\n            you read the entire contents of the response such as when\n            `preload_content=True`). This is useful if you're not preloading\n            the response's content immediately. You will need to call\n            ``r.release_conn()`` on the response ``r`` to return the connection\n            back into the pool. If None, it takes the value of ``preload_content``\n            which defaults to ``True``.\n    \n        :param bool chunked:\n            If True, urllib3 will send the body using chunked transfer\n            encoding. Otherwise, urllib3 will send the body using the standard\n            content-length form. Defaults to False.\n    \n        :param int body_pos:\n            Position to seek to in file-like body in the event of a retry or\n            redirect. Typically this won't need to be set because urllib3 will\n            auto-populate the value when needed.\n        \"\"\"\n        parsed_url = parse_url(url)\n        destination_scheme = parsed_url.scheme\n    \n        if headers is None:\n            headers = self.headers\n    \n        if not isinstance(retries, Retry):\n            retries = Retry.from_int(retries, redirect=redirect, default=self.retries)\n    \n        if release_conn is None:\n            release_conn = preload_content\n    \n        # Check host\n        if assert_same_host and not self.is_same_host(url):\n            raise HostChangedError(self, url, retries)\n    \n        # Ensure that the URL we're connecting to is properly encoded\n        if url.startswith(\"/\"):\n            url = to_str(_encode_target(url))\n        else:\n            url = to_str(parsed_url.url)\n    \n        conn = None\n    \n        # Track whether `conn` needs to be released before\n        # returning/raising/recursing. Update this variable if necessary, and\n        # leave `release_conn` constant throughout the function. That way, if\n        # the function recurses, the original value of `release_conn` will be\n        # passed down into the recursive call, and its value will be respected.\n        #\n        # See issue #651 [1] for details.\n        #\n        # [1] <https://github.com/urllib3/urllib3/issues/651>\n        release_this_conn = release_conn\n    \n        http_tunnel_required = connection_requires_http_tunnel(\n            self.proxy, self.proxy_config, destination_scheme\n        )\n    \n        # Merge the proxy headers. Only done when not using HTTP CONNECT. We\n        # have to copy the headers dict so we can safely change it without those\n        # changes being reflected in anyone else's copy.\n        if not http_tunnel_required:\n            headers = headers.copy()  # type: ignore[attr-defined]\n            headers.update(self.proxy_headers)  # type: ignore[union-attr]\n    \n        # Must keep the exception bound to a separate variable or else Python 3\n        # complains about UnboundLocalError.\n        err = None\n    \n        # Keep track of whether we cleanly exited the except block. This\n        # ensures we do proper cleanup in finally.\n        clean_exit = False\n    \n        # Rewind body position, if needed. Record current position\n        # for future rewinds in the event of a redirect/retry.\n        body_pos = set_file_position(body, body_pos)\n    \n        try:\n            # Request a connection from the queue.\n            timeout_obj = self._get_timeout(timeout)\n            conn = self._get_conn(timeout=pool_timeout)\n    \n            conn.timeout = timeout_obj.connect_timeout  # type: ignore[assignment]\n    \n            # Is this a closed/new connection that requires CONNECT tunnelling?\n            if self.proxy is not None and http_tunnel_required and conn.is_closed:\n                try:\n                    self._prepare_proxy(conn)\n                except (BaseSSLError, OSError, SocketTimeout) as e:\n                    self._raise_timeout(\n                        err=e, url=self.proxy.url, timeout_value=conn.timeout\n                    )\n                    raise\n    \n            # If we're going to release the connection in ``finally:``, then\n            # the response doesn't need to know about the connection. Otherwise\n            # it will also try to release it and we'll have a double-release\n            # mess.\n            response_conn = conn if not release_conn else None\n    \n            # Make the request on the HTTPConnection object\n>           response = self._make_request(\n                conn,\n                method,\n                url,\n                timeout=timeout_obj,\n                body=body,\n                headers=headers,\n                chunked=chunked,\n                retries=retries,\n                response_conn=response_conn,\n                preload_content=preload_content,\n                decode_content=decode_content,\n                **response_kw,\n            )\n\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:787: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:493: in _make_request\n    conn.request(\n.venv/lib/python3.12/site-packages/urllib3/connection.py:445: in request\n    self.endheaders()\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1314: in endheaders\n    self._send_output(message_body, encode_chunked=encode_chunked)\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1074: in _send_output\n    self.send(msg)\n/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/client.py:1018: in send\n    self.connect()\n.venv/lib/python3.12/site-packages/urllib3/connection.py:276: in connect\n    self.sock = self._new_conn()\n                ^^^^^^^^^^^^^^^^\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\nself = <urllib3.connection.HTTPConnection object at 0x107fe2f00>\n\n    def _new_conn(self) -> socket.socket:\n        \"\"\"Establish a socket connection and set nodelay settings on it.\n    \n        :return: New socket connection.\n        \"\"\"\n        try:\n            sock = connection.create_connection(\n                (self._dns_host, self.port),\n                self.timeout,\n                source_address=self.source_address,\n                socket_options=self.socket_options,\n            )\n        except socket.gaierror as e:\n            raise NameResolutionError(self.host, self, e) from e\n        except SocketTimeout as e:\n            raise ConnectTimeoutError(\n                self,\n                f\"Connection to {self.host} timed out. (connect timeout={self.timeout})\",\n            ) from e\n    \n        except OSError as e:\n>           raise NewConnectionError(\n                self, f\"Failed to establish a new connection: {e}\"\n            ) from e\nE           urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x107fe2f00>: Failed to establish a new connection: [Errno 61] Connection refused\n\n.venv/lib/python3.12/site-packages/urllib3/connection.py:213: NewConnectionError\n\nThe above exception was the direct cause of the following exception:\n\n    @allure.tag(\"web\")\n    @allure.title(\"Успешная регистрация студента с полными данными\")\n    @allure.label(\"owner\", \"Max Razzhivin\")\n    @allure.feature(\"Форма для заполнения студентом своих данных\")\n    @allure.story(\"Форма корректно заполняется и результаты корректные\")\n    @allure.link(\"https://demoqa.com/automation-practice-form\", \"Test\")\n    @allure.severity(Severity.CRITICAL)\n    def test_complete_do():\n        # GIVEN:\n        registration_page = RegistrationPage()\n        registration_page.open()\n    \n        # WHEN\n        registration_page.fill_first_name('Max')\n        registration_page.fill_last_name('Razzhivin')\n        registration_page.fill_email('max.nvo06@gmail.com')\n        registration_page.fill_gender('Male')\n    \n        registration_page.fill_user_number('9094618666')\n    \n        # Выбираем дату рождения:\n        registration_page.fill_date_of_birth('1989', 'Apr', '06')\n    \n        # Предметы\n        registration_page.fill_subjects('Computer Science')\n    \n    \n        # Хобби\n        registration_page.fill_hobbies('Sports', 'Reading')\n    \n        registration_page.fill_image()\n    \n        registration_page.fill_address('somewhere in galaxy')\n    \n        registration_page.fill_state('NCR')\n        registration_page.fill_city('Delhi')\n    \n        registration_page.submit()\n    \n        # THEN\n>       registration_page.should_registered_user_with(\"Max Razzhivin\",\n                \"max.nvo06@gmail.com\",\n                'Male',\n                '9094618666',\n                '06 April,1989',\n                'Computer Science',\n                'Sports, Reading',\n                'picta.png',\n                'somewhere in galaxy',\n                'NCR Delhi',)\n\ntests/test_page_object.py:46: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \npages/registration_page.py:121: in should_registered_user_with\n    browser.element(\".table\").all('td').even.should(\n.venv/lib/python3.12/site-packages/selene/core/entity.py:132: in should\n    self.wait.for_(condition)\n.venv/lib/python3.12/site-packages/selene/core/wait.py:150: in for_\n    return self._decorator(self)(logic)(fn)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/wait.py:148: in logic\n    raise self._hook_failure(failure)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                             ^^^^\n.venv/lib/python3.12/site-packages/selene/core/configuration.py:1411: in save_and_log_screenshot\n    path = self._save_screenshot_strategy(self)  # type: ignore\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/core/configuration.py:1238: in <lambda>\n    lambda config, path=None: fp.thread(  # type: ignore\n.venv/lib/python3.12/site-packages/selene/common/fp.py:83: in thread\n    return pipe(*functions)(arg)\n           ^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n.venv/lib/python3.12/site-packages/selene/common/fp.py:73: in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                             ^^^^\n.venv/lib/python3.12/site-packages/selene/core/configuration.py:1269: in <lambda>\n    lambda path: (path if config.driver.get_screenshot_as_file(path) else None),\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:996: in get_screenshot_as_file\n    png = self.get_screenshot_as_png()\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:1030: in get_screenshot_as_png\n    return b64decode(self.get_screenshot_as_base64().encode(\"ascii\"))\n                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:1040: in get_screenshot_as_base64\n    return self.execute(Command.SCREENSHOT)[\"value\"]\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py:451: in execute\n    response = cast(RemoteConnection, self.command_executor).execute(driver_command, params)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py:404: in execute\n    return self._request(command_info[0], url, body=data)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py:428: in _request\n    response = self._conn.request(method, url, body=body, headers=headers, timeout=self._client_config.timeout)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/_request_methods.py:135: in request\n    return self.request_encode_url(\n.venv/lib/python3.12/site-packages/urllib3/_request_methods.py:182: in request_encode_url\n    return self.urlopen(method, url, **extra_kw)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/poolmanager.py:443: in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:871: in urlopen\n    return self.urlopen(\n.venv/lib/python3.12/site-packages/urllib3/connectionpool.py:841: in urlopen\n    retries = retries.increment(\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\nself = Retry(total=0, connect=None, read=None, redirect=None, status=None), method = 'GET', url = '/session/94e97799e125329f558cf44c1e7b7b38/screenshot'\nresponse = None\nerror = NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2f00>: Failed to establish a new connection: [Errno 61] Connection refused')\n_pool = <urllib3.connectionpool.HTTPConnectionPool object at 0x102f72300>, _stacktrace = <traceback object at 0x107ffe3c0>\n\n    def increment(\n        self,\n        method: str | None = None,\n        url: str | None = None,\n        response: BaseHTTPResponse | None = None,\n        error: Exception | None = None,\n        _pool: ConnectionPool | None = None,\n        _stacktrace: TracebackType | None = None,\n    ) -> Self:\n        \"\"\"Return a new Retry object with incremented retry counters.\n    \n        :param response: A response object, or None, if the server did not\n            return a response.\n        :type response: :class:`~urllib3.response.BaseHTTPResponse`\n        :param Exception error: An error encountered during the request, or\n            None if the response was received successfully.\n    \n        :return: A new ``Retry`` object.\n        \"\"\"\n        if self.total is False and error:\n            # Disabled, indicate to re-raise the error.\n            raise reraise(type(error), error, _stacktrace)\n    \n        total = self.total\n        if total is not None:\n            total -= 1\n    \n        connect = self.connect\n        read = self.read\n        redirect = self.redirect\n        status_count = self.status\n        other = self.other\n        cause = \"unknown\"\n        status = None\n        redirect_location = None\n    \n        if error and self._is_connection_error(error):\n            # Connect retry?\n            if connect is False:\n                raise reraise(type(error), error, _stacktrace)\n            elif connect is not None:\n                connect -= 1\n    \n        elif error and self._is_read_error(error):\n            # Read retry?\n            if read is False or method is None or not self._is_method_retryable(method):\n                raise reraise(type(error), error, _stacktrace)\n            elif read is not None:\n                read -= 1\n    \n        elif error:\n            # Other retry?\n            if other is not None:\n                other -= 1\n    \n        elif response and response.get_redirect_location():\n            # Redirect retry?\n            if redirect is not None:\n                redirect -= 1\n            cause = \"too many redirects\"\n            response_redirect_location = response.get_redirect_location()\n            if response_redirect_location:\n                redirect_location = response_redirect_location\n            status = response.status\n    \n        else:\n            # Incrementing because of a server error like a 500 in\n            # status_forcelist and the given method is in the allowed_methods\n            cause = ResponseError.GENERIC_ERROR\n            if response and response.status:\n                if status_count is not None:\n                    status_count -= 1\n                cause = ResponseError.SPECIFIC_ERROR.format(status_code=response.status)\n                status = response.status\n    \n        history = self.history + (\n            RequestHistory(method, url, error, status, redirect_location),\n        )\n    \n        new_retry = self.new(\n            total=total,\n            connect=connect,\n            read=read,\n            redirect=redirect,\n            status=status_count,\n            other=other,\n            history=history,\n        )\n    \n        if new_retry.is_exhausted():\n            reason = error or ResponseError(cause)\n>           raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]\n            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\nE           urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=59269): Max retries exceeded with url: /session/94e97799e125329f558cf44c1e7b7b38/screenshot (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2f00>: Failed to establish a new connection: [Errno 61] Connection refused'))\n\n.venv/lib/python3.12/site-packages/urllib3/util/retry.py:519: MaxRetryError"}, "steps": [{"name": "Открываем браузер и удаляем рекламу", "status": "passed", "start": 1752685838477, "stop": 1752685842651}, {"name": "Заполняем имя: 'Max'", "status": "passed", "parameters": [{"name": "value", "value": "'Max'"}], "start": 1752685842651, "stop": 1752685842722}, {"name": "Заполняем фамилию: 'Razzhivin'", "status": "passed", "parameters": [{"name": "value", "value": "'Razzhivin'"}], "start": 1752685842722, "stop": 1752685842739}, {"name": "Заполняем почту: 'max.nvo06@gmail.com'", "status": "passed", "parameters": [{"name": "value", "value": "'max.nvo06@gmail.com'"}], "start": 1752685842739, "stop": 1752685842756}, {"name": "Выбираем пол: 'Male'", "status": "passed", "parameters": [{"name": "value", "value": "'Male'"}], "start": 1752685842756, "stop": 1752685842793}, {"name": "Заполняем номер телефона: '9094618666'", "status": "passed", "parameters": [{"name": "value", "value": "'9094618666'"}], "start": 1752685842793, "stop": 1752685842811}, {"name": "Заполняем дату рождения: '1989' 'Apr' '06'", "status": "passed", "parameters": [{"name": "year", "value": "'1989'"}, {"name": "month", "value": "'Apr'"}, {"name": "day", "value": "'06'"}], "start": 1752685842812, "stop": 1752685842916}, {"name": "Выбираем предметы: 'Computer Science'", "status": "passed", "parameters": [{"name": "value", "value": "'Computer Science'"}], "start": 1752685842916, "stop": 1752685842961}, {"name": "Выбираем хобби: 'Sports' 'Reading'", "status": "passed", "parameters": [{"name": "value1", "value": "'Sports'"}, {"name": "value2", "value": "'Reading'"}], "start": 1752685842961, "stop": 1752685846062}, {"name": "Загружаем картинку", "status": "passed", "start": 1752685846062, "stop": 1752685846089}, {"name": "Заполняем адрес: 'somewhere in galaxy'", "status": "passed", "parameters": [{"name": "value", "value": "'somewhere in galaxy'"}], "start": 1752685846089, "stop": 1752685846122}, {"name": "Выбираем штат: 'NCR'", "status": "passed", "parameters": [{"name": "value", "value": "'NCR'"}], "start": 1752685846122, "stop": 1752685846180}, {"name": "Выбираем город: 'Delhi'", "status": "passed", "parameters": [{"name": "value", "value": "'Delhi'"}], "start": 1752685846180, "stop": 1752685846221}, {"name": "Нажимаем кнопку \"submit\"", "status": "passed", "start": 1752685846221, "stop": 1752685846229}, {"name": "Проверяем результаты формы", "status": "broken", "statusDetails": {"message": "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=59269): Max retries exceeded with url: /session/94e97799e125329f558cf44c1e7b7b38/screenshot (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x107fe2f00>: Failed to establish a new connection: [Errno 61] Connection refused'))\n", "trace": "  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/allure_commons/_allure.py\", line 202, in impl\n    return func(*a, **kw)\n           ^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/pages/registration_page.py\", line 121, in should_registered_user_with\n    browser.element(\".table\").all('td').even.should(\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/entity.py\", line 132, in should\n    self.wait.for_(condition)\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/wait.py\", line 150, in for_\n    return self._decorator(self)(logic)(fn)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/wait.py\", line 148, in logic\n    raise self._hook_failure(failure)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/common/fp.py\", line 73, in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                             ^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/configuration.py\", line 1411, in save_and_log_screenshot\n    path = self._save_screenshot_strategy(self)  # type: ignore\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/configuration.py\", line 1238, in <lambda>\n    lambda config, path=None: fp.thread(  # type: ignore\n                              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/common/fp.py\", line 83, in thread\n    return pipe(*functions)(arg)\n           ^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/common/fp.py\", line 73, in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/common/fp.py\", line 73, in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/common/fp.py\", line 73, in <lambda>\n    lambda f, g: lambda x: f(g(x)) if g else f(x),\n                           ^^^^^^^\n  [Previous line repeated 2 more times]\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selene/core/configuration.py\", line 1269, in <lambda>\n    lambda path: (path if config.driver.get_screenshot_as_file(path) else None),\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py\", line 996, in get_screenshot_as_file\n    png = self.get_screenshot_as_png()\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py\", line 1030, in get_screenshot_as_png\n    return b64decode(self.get_screenshot_as_base64().encode(\"ascii\"))\n                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py\", line 1040, in get_screenshot_as_base64\n    return self.execute(Command.SCREENSHOT)[\"value\"]\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/webdriver.py\", line 451, in execute\n    response = cast(RemoteConnection, self.command_executor).execute(driver_command, params)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py\", line 404, in execute\n    return self._request(command_info[0], url, body=data)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/selenium/webdriver/remote/remote_connection.py\", line 428, in _request\n    response = self._conn.request(method, url, body=body, headers=headers, timeout=self._client_config.timeout)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/_request_methods.py\", line 135, in request\n    return self.request_encode_url(\n           ^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/_request_methods.py\", line 182, in request_encode_url\n    return self.urlopen(method, url, **extra_kw)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/poolmanager.py\", line 443, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/connectionpool.py\", line 871, in urlopen\n    return self.urlopen(\n           ^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/connectionpool.py\", line 871, in urlopen\n    return self.urlopen(\n           ^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/connectionpool.py\", line 871, in urlopen\n    return self.urlopen(\n           ^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/connectionpool.py\", line 841, in urlopen\n    retries = retries.increment(\n              ^^^^^^^^^^^^^^^^^^\n  File \"/Users/maxnovo/Desktop/qa.guru/18.08 Page_Object/.venv/lib/python3.12/site-packages/urllib3/util/retry.py\", line 519, in increment\n    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n"}, "parameters": [{"name": "full_name", "value": "'Max Razzhivin'"}, {"name": "email", "value": "'max.nvo06@gmail.com'"}, {"name": "gender", "value": "'Male'"}, {"name": "mobile", "value": "'9094618666'"}, {"name": "date_of_birth", "value": "'06 April,1989'"}, {"name": "subject", "value": "'Computer Science'"}, {"name": "hobbies", "value": "'Sports, Reading'"}, {"name": "image_name", "value": "'picta.png'"}, {"name": "address", "value": "'somewhere in galaxy'"}, {"name": "state_city", "value": "'NCR Delhi'"}], "start": 1752685846229, "stop": 1752685850231}], "attachments": [{"name": "log", "source": "3556ed95-ee9e-43f8-8f71-838674a37705-attachment.txt", "type": "text/plain"}, {"name": "stderr", "source": "a3b7519d-f6ec-41f6-acb3-9e6476c116ce-attachment.txt", "type": "text/plain"}], "start": 1752685838477, "stop": 1752685850237, "uuid": "9bbb825a-6681-44b4-9aaa-be14c6d28f98", "historyId": "f8638fbca7d3fb3e8502a78a75c9b25b", "testCaseId": "f8638fbca7d3fb3e8502a78a75c9b25b", "fullName": "tests.test_page_object#test_complete_do", "labels": [{"name": "feature", "value": "Форма для заполнения студентом своих данных"}, {"name": "severity", "value": "critical"}, {"name": "story", "value": "Форма корректно заполняется и результаты корректные"}, {"name": "owner", "value": "Max Razzhivin"}, {"name": "tag", "value": "web"}, {"name": "parentSuite", "value": "tests"}, {"name": "suite", "value": "test_page_object"}, {"name": "host", "value": "MacBook-Pro.local"}, {"name": "thread", "value": "27124-MainThread"}, {"name": "framework", "value": "pytest"}, {"name": "language", "value": "cpython3"}, {"name": "package", "value": "tests.test_page_object"}], "links": [{"type": "Test", "url": "https://demoqa.com/automation-practice-form", "name": "https://demoqa.com/automation-practice-form"}]}